### PR TITLE
[WIP] Mangle imported declarations using their C names

### PR DESCRIPTION
**Pull request by jrose-apple**
This makes them consistent no matter what shenanigans are pulled by the importer, particularly NS_ENUM vs. NS_OPTIONS and NS_SWIFT_NAME.

The 'NSErrorDomain' API note *nearly* works with this, but the synthesized error struct is still mangled as a Swift declaration, which means it's not rename-stable. Fortunately, these are rare to appear at API boundaries.

The one place where this still falls down is NS_STRING_ENUM: when this is applied, a typedef is imported as a unique struct, but without it it's just a typealias for the underlying type. I don't have a solution for this yet.

This PR pulls in #8849 and a number of other little cleanups; the important part is the last two commits. It still needs more tests, and may not even be the direction we want to go.

rdar://problem/31616162

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@swift-ci Please clean test Linux

File: stdlib/public/Reflection/TypeRef.cpp [hunk 1]
```diff
@@ -72,6 +72,8 @@ class PrintTypeRef : public TypeRefVisitor<PrintTypeRef, void> {
       printHeader("enum");
     else if (N->isClass())
       printHeader("class");
+    else if (N->isAlias())
```
Comment (slavapestov) [inline stdlib/public/Reflection/TypeRef.cpp]: This needs tests

Comment (jrose-apple) [inline stdlib/public/Reflection/TypeRef.cpp]: It's tested!…in one place. :-)

**Comment by jrose-apple** [comment]
@swift-ci Please test Linux

File: include/swift/Strings.h [hunk 2]
```diff
@@ -39,9 +39,9 @@ namespace swift {
   static const char LLDB_EXPRESSIONS_MODULE_NAME_PREFIX[] = "__lldb_expr_";
 
   /// The name of the fake module used to hold imported Objective-C things.
-  static const char MANGLING_MODULE_OBJC[] = "__ObjC";
-  /// The name of the fake module used to hold imported C things.
-  static const char MANGLING_MODULE_C[] = "__C";
+  static const char MANGLING_MODULE_OBJC[] = "__C";
```
Comment (jckarter) [inline include/swift/Strings.h]: Do we have downstream tools that depend on the `__C` and `__ObjC` demanglings being as they are, say lldb or Instruments?

File: lib/IRGen/GenProto.cpp [hunk 3]
```diff
@@ -1734,7 +1735,7 @@ void IRGenModule::emitSILWitnessTable(SILWitnessTable *wt) {
     if (ClassDecl *clas = dyn_cast<ClassDecl>(Nominal)) {
       if (clas->isForeign())
         getAddrOfForeignTypeMetadataCandidate(conformingType);
-    } else if (Nominal->hasClangNode()) {
+    } else if (isa<ClangModuleUnit>(Nominal->getModuleScopeContext())) {
```
Comment (jckarter) [inline lib/IRGen/GenProto.cpp]: Should `hasClangNode` be replaced by a `isInClangModuleContext` method on ValueDecl? If it's the wrong thing to reach for for this kind of query, it seems like we should make the right thing at least as obvious.

File: include/swift/Strings.h (hunk 2 above)
Comment (jrose-apple) [inline include/swift/Strings.h]: LLDB does not, but I did forget to check the other Apple-internal tools. Thanks for the reminder.

File: lib/IRGen/GenProto.cpp (hunk 3 above)
Comment (jrose-apple) [inline lib/IRGen/GenProto.cpp]: Yeah, that's not a bad idea.

File: include/swift/Strings.h (hunk 2 above)
Comment (jrose-apple) [inline include/swift/Strings.h]: Seems to be okay.

**Comment by jrose-apple** [comment]
To be picked up in the future; for now we're going with the much similar #8987.

**Pull request closed without merging**